File: 2023/day1/day1.ex
defmodule StringUtil do
  def calib_val(str) do
    first_digit(str) + last_digit(str)
  end

  def first_digit(str) when is_binary(str) do
    ret_digit(str)
  end

  def last_digit(str) when is_binary(str) do
    ret_digit(String.reverse(str))
  end

  defp ret_digit(<<c :: utf8, _rest :: binary>>) when c in ?0..?9 do
    c
  end

  defp ret_digit(<<_c :: utf8, rest :: binary>>) do
    ret_digit(rest)
  end
end

defmodule Mathieu do
  def calib_val(line) do
    first_num = String.sub("^\\D+(\\d+).*", "\\1", line)
    last_num = String.reverse(line) |> String.sub("^\\D+(\\d+).*", "\\1")
    first_num + last_num
  end
end

{:ok, contents} = File.read("seed1.txt")
lines = String.split(contents, "\n", trim: true)
values = Enum.map(lines, &StringUtil.calib_val/1)
# sum = contents |>  |>  |>
IO.inspect(StringUtil.last_digit("asdfghjkfvjr28vw2893asdf7ghjkl"))

IO.puts Enum.sum(values)

File: 2023/day1/day1.rb
require 'byebug'

Numbers = %w[one two three four five six seven eight nine]

def fix_data(data)
  data.gsub!(/oneight/, "oneeight")
  data.gsub!(/threeight/, "threeeight")
  data.gsub!(/fiveight/, "fiveeight")
  data.gsub!(/nineight/, "nineeight")
  data.gsub!(/twone/, "twoone")
  data.gsub!(/sevenine/, "sevennine")
  data.gsub!(/eightwo/, "eighttwo")
end

def word_to_digit(word)
  Numbers.index(word)+1
end

def replace_first_and_last(line)
  forward_match = line.match(%r[(#{Numbers.join('|')})])
  reverse_match = line.reverse.match(%r[(#{Numbers.map(&:reverse).join('|')})])
  if forward_match
      line.gsub!(forward_match[0], word_to_digit(forward_match[0]).to_s)
  end
  if reverse_match
      bingo = reverse_match[0].reverse
      puts bingo
      line.gsub!(bingo, word_to_digit(bingo).to_s)
  end
end

def replace_worded_numbers(line)
  while forward_match = line.match(%r[(#{Numbers.join('|')})])
    if forward_match
      line.gsub!(forward_match[0], word_to_digit(forward_match[0]).to_s)
    end
  end
end

def calib_val(line)
    puts "before> #{line}"
    replace_worded_numbers(line)
    puts "after>  #{line}"
    first_number = line.match(/\d+/)[0][0].to_i
    last_number = line.reverse.match(/\d+/)[0][0].to_i
    answer = (10*first_number) + last_number
    puts "answer> #{answer}"
    answer
end

data = File.read("day1-seed1.txt")
fix_data(data)
contents = data.split("\n").compact

puts contents.sum { |line| calib_val(line) }
